### ISSUE TITLE
Feature Request: Support for Dynamic Bias Tensor in FlexAttention Without Recompilation

### ISSUE THREAD

**Issue by pengzhangzhi** [pytorch-labs/attention-gym#123]
Hi, 

FlexAttention currently requires recompilation whenever the score_mod function changes. This works well for static attention modifications but presents challenges when the modification involves a dynamically computed bias tensor that varies across batches in training.

I am working on a use case where I need to add a bias term of shape [B, H, S, S] to the attention scores before applying the softmax operation. However, since this bias is computed as an intermediate result in the forward pass, embedding it in the score_mod function forces a full kernel recompilation on every forward pass, significantly impacting performance.

Would it be possible to extend FlexAttention to allow dynamically varying bias tensors without requiring recompilation?

Current Issue
	•	Problem: FlexAttention’s design requires a static function for score_mod, meaning any modifications passed as closures lead to kernel recompilation whenever the function (or its captured tensors) changes.
	•	Why This is a Problem: In many real-world applications (e.g., Transformer models with learned attention biases), the bias is dynamically computed during the forward pass. Since recompilation is expensive, this makes FlexAttention impractical for training scenarios where the bias differs across batches.

Proposed Solution

A potential way to address this could be:
	1.	Allowing a bias argument directly in flex_attention(...) that accepts a tensor of shape [B, H, S, S] and integrates it into the fused kernel computation.
	2.	Distinguishing between static modifications (handled via score_mod) and dynamic biases (passed as a tensor input) so that frequent recompilation is avoided.
	3.	Enabling optional dynamic bias support using CUDA/Triton shared memory or efficient in-kernel broadcasting to minimize performance overhead.

Would love to hear thoughts from the maintainers on the feasibility of this and whether such an extension aligns with the design goals of FlexAttention!

**Comment by Bhazantri** [comment]
Design parameterized kernels that accept a set of parameters, such as bias tensors and scaling factors, which can be dynamically adjusted without requiring recompilation, leveraging Triton’s template-based kernel generation or CUDA’s dynamic parallelism. Implementing a lazy recompilation mechanism ensures that the kernel is only recompiled when the structure of the score_mod function changes, such as the introduction of new operations, while changes to the bias tensor’s values or shape do not trigger recompilation. Kernel fusion techniques can be employed to combine the attention computation and bias addition into a single kernel, reducing overhead and optimizing memory access patterns. Additionally, an adaptive kernel selection system can be developed to choose the most efficient kernel variant at runtime based on input characteristics like batch size, sequence length, and bias tensor shape, potentially using a decision tree or machine learning model trained on kernel performance data. Further research-level optimizations can include leveraging sparse bias tensors with sparse matrix operations to reduce computation and memory overhead, quantizing the bias tensor to lower precision (e.g., FP16 or INT8) to enhance memory bandwidth and computation speed, and employing hierarchical attention to decompose the computation into smaller blocks, applying the bias at the block level to improve locality and parallelism. These advancements would significantly enhance the flexibility and performance of FlexAttention in handling dynamic attention biases during training.

Implimentation: 

import triton
import triton.language as tl

@triton.jit
def flex_attention_kernel(
    query_ptr, key_ptr, value_ptr, output_ptr, bias_ptr,
    B, H, S, D,
    score_mod=None,  # Static function
    BLOCK_SIZE: tl.constexpr,
):
    pid_b = tl.program_id(axis=0)
    pid_h = tl.program_id(axis=1)
    pid_s = tl.program_id(axis=2)

    # Load query, key, and value blocks
    query = tl.load(query_ptr + pid_b * H * S * D + pid_h * S * D + pid_s * D)
    key = tl.load(key_ptr + pid_b * H * S * D + pid_h * S * D + pid_s * D)
    value = tl.load(value_ptr + pid_b * H * S * D + pid_h * S * D + pid_s * D)

    # Compute attention scores
    scores = tl.dot(query, key)
    if bias_ptr is not None:
        bias = tl.load(bias_ptr + pid_b * H * S * S + pid_h * S * S + pid_s * S)
        scores += bias

    # Apply score_mod if provided
    if score_mod is not None:
        scores = score_mod(scores)

    # Softmax and weighted sum
    scores = tl.softmax(scores)
    output = tl.dot(scores, value)
    tl.store(output_ptr + pid_b * H * S * D + pid_h * S * D + pid_s * D, output)

def flex_attention(query, key, value, score_mod=None, bias=None):
    B, H, S, D = query.shape
    output = torch.empty_like(query)
    grid = (B, H, S)
    flex_attention_kernel[grid](
        query, key, value, output, bias,
        B, H, S, D,
        score_mod=score_mod,
        BLOCK_SIZE=128,
    )
    return output

**Comment by drisspg** [comment]
If you are adding a ` [B, H, S, S]` bias tensor then you honestly shouldn't be using FlexAttention and torch.nn.functional.scaled_dot_product_attention is both a better API and faster
